There are some unit test failures.

```
[wtam@localhost pipelines]$ go test ./...
Adding .git to https://github.com/org/gitops
Adding .git to https://github.com/org/app
Adding .git to test/repo
Adding .git to test/repo
Adding .git to test/repo
  
Checking dependencies

--- FAIL: TestDependenciesWithNothingInstalled (0.00s)
    bootstrap_test.go:221: Message mismatch: got 
        Checking if Sealed Secrets is installed with the default configuration[Please install Sealed Secrets from https://github.com/bitnami-labs/sealed-secrets/releases]
        Checking if ArgoCD Operator is installed with the default configuration[Please install ArgoCD operator from OperatorHub, with an ArgoCD resource called 'argocd']
        Checking if OpenShift Pipelines Operator is installed with the default configuration[Please install OpenShift Pipelines operator from OperatorHub], want 
        Checking if Sealed Secrets is installed with the default configuration[Please install Sealed Secrets from https://github.com/bitnami-labs/sealed-secrets/releases]
        Checking if ArgoCD Operator is installed with the default configuration[Please install ArgoCD operator from OperatorHub]
        Checking if OpenShift Pipelines Operator is installed with the default configuration[Please install OpenShift Pipelines operator from OperatorHub]
  
Checking dependencies

  
Checking dependencies

--- FAIL: TestDependenciesWithNoArgoCD (0.00s)
    bootstrap_test.go:259: Message mismatch: got 
        Checking if Sealed Secrets is installed with the default configuration
        Checking if ArgoCD Operator is installed with the default configuration[Please install ArgoCD operator from OperatorHub, with an ArgoCD resource called 'argocd']
        Checking if OpenShift Pipelines Operator is installed with the default configuration, want 
        Checking if Sealed Secrets is installed with the default configuration
        Checking if ArgoCD Operator is installed with the default configuration[Please install ArgoCD operator from OperatorHub]
        Checking if OpenShift Pipelines Operator is installed with the default configuration
  
Checking dependencies

FAIL
FAIL    github.com/openshift/odo/pkg/odo/cli/pipelines  0.019s
ok      github.com/openshift/odo/pkg/odo/cli/pipelines/environment      (cached)
ok      github.com/openshift/odo/pkg/odo/cli/pipelines/service  (cached)
ok      github.com/openshift/odo/pkg/odo/cli/pipelines/ui       (cached)
ok      github.com/openshift/odo/pkg/odo/cli/pipelines/utility  (cached)
ok      github.com/openshift/odo/pkg/odo/cli/pipelines/webhook  (cached)
FAIL
```

--- a/pkg/odo/cli/pipelines/bootstrap.go
+++ b/pkg/odo/cli/pipelines/bootstrap.go
@@ -29,8 +29,6 @@ const (
 	sealedSecretsName   = "sealed-secrets-controller"
 	sealedSecretsNS     = "kube-system"
 	argoCDNS            = "argocd"
-	argoCDOperatorName  = "argocd-operator"
-	argoCDServerName    = "argocd-server"
 	pipelinesOperatorNS = "openshift-operators"
 )
 
@@ -84,7 +82,6 @@ func NewBootstrapParameters() *BootstrapParameters {
 // If the prefix provided doesn't have a "-" then one is added, this makes the
 // generated environment names nicer to read.
 func (io *BootstrapParameters) Complete(name string, cmd *cobra.Command, args []string) error {
-
 	clientSet, err := namespaces.GetClientSet()
 	if err != nil {
 		return err
@@ -141,8 +138,7 @@ func initiateInteractiveMode(io *BootstrapParameters) error {
 		io.SealedSecretsService.Name = ui.EnterSealedSecretService(&io.SealedSecretsService)
 
 	}
-	io.GitOpsRepoURL = ui.EnterGitRepo()
-	io.GitOpsRepoURL = utility.AddGitSuffixIfNecessary(io.GitOpsRepoURL)
+	io.GitOpsRepoURL = utility.AddGitSuffixIfNecessary(ui.EnterGitRepo())
 	if !isKnownDriver(io.GitOpsRepoURL) {
 		io.PrivateRepoDriver = ui.SelectPrivateRepoDriver()
 		host, err := hostFromURL(io.GitOpsRepoURL)
@@ -157,15 +153,21 @@ func initiateInteractiveMode(io *BootstrapParameters) error {
 		io.InternalRegistryHostname = ui.EnterInternalRegistry()
 		io.ImageRepo = ui.EnterImageRepoInternalRegistry()
 	} else {
-		io.DockerConfigJSONFilename = ui.EnterDockercfg()
 		io.ImageRepo = ui.EnterImageRepoExternalRepository()
+		io.DockerConfigJSONFilename = ui.EnterDockercfg()
 	}
 	io.GitOpsWebhookSecret = ui.EnterGitWebhookSecret()
 	io.ServiceRepoURL = ui.EnterServiceRepoURL()
+	if ui.IsPrivateRepo() {
+		io.GitHostAccessToken = ui.EnterGitHostAccessToken(io.ServiceRepoURL)
+	}
 	io.ServiceWebhookSecret = ui.EnterServiceWebhookSecret()
 	commitStatusTrackerCheck := ui.SelectOptionCommitStatusTracker()
 	if commitStatusTrackerCheck == "yes" {
-		io.StatusTrackerAccessToken = ui.EnterStatusTrackerAccessToken(io.ServiceRepoURL)
+		io.CommitStatusTracker = true
+		if io.GitHostAccessToken == "" {
+			io.GitHostAccessToken = ui.EnterGitHostAccessToken(io.ServiceRepoURL)
+		}
 	}
 	io.Prefix = ui.EnterPrefix()
 	io.OutputPath = ui.EnterOutputPath()
@@ -190,7 +192,7 @@ func checkBootstrapDependencies(io *BootstrapParameters, kubeClient kubernetes.I
 
 	spinner.Start("Checking if ArgoCD Operator is installed with the default configuration", false)
 	err = client.CheckIfArgoCDExists(argoCDNS)
-	setSpinnerStatus(spinner, "Please install ArgoCD operator from OperatorHub", err)
+	setSpinnerStatus(spinner, "Please install ArgoCD operator from OperatorHub, with an ArgoCD resource called 'argocd'", err)
 	if err != nil {
 		if !errors.IsNotFound(err) {
 			return clusterErr(err.Error())
@@ -232,7 +234,7 @@ func (io *BootstrapParameters) Validate() error {
 		return fmt.Errorf("failed to parse url %s: %w", io.GitOpsRepoURL, err)
 	}
 
-	// TODO: this won't work with GitLab as the repo can have more path elements.
+	// TODO: this may not work with GitLab as the repo can have more path elements.
 	if len(utility.RemoveEmptyStrings(strings.Split(gr.Path, "/"))) != 2 {
 		return fmt.Errorf("repo must be org/repo: %s", strings.Trim(gr.Path, ".git"))
 	}
@@ -282,11 +284,12 @@ func NewCmdBootstrap(name, fullName string) *cobra.Command {
 	bootstrapCmd.Flags().StringVar(&o.ImageRepo, "image-repo", "", "Image repository of the form <registry>/<username>/<repository> or <project>/<app> which is used to push newly built images")
 	bootstrapCmd.Flags().StringVar(&o.SealedSecretsService.Namespace, "sealed-secrets-ns", "kube-system", "Namespace in which the Sealed Secrets operator is installed, automatically generated secrets are encrypted with this operator")
 	bootstrapCmd.Flags().StringVar(&o.SealedSecretsService.Name, "sealed-secrets-svc", "sealed-secrets-controller", "Name of the Sealed Secrets Services that encrypts secrets")
-	bootstrapCmd.Flags().StringVar(&o.StatusTrackerAccessToken, "status-tracker-access-token", "", "Used to authenticate requests to push commit-statuses to your Git hosting service")
+	bootstrapCmd.Flags().StringVar(&o.GitHostAccessToken, "git-host-access-token", "", "Used to authenticate repository clones, and commit-status notifications (if enabled)")
 	bootstrapCmd.Flags().BoolVar(&o.Overwrite, "overwrite", false, "Overwrites previously existing GitOps configuration (if any)")
 	bootstrapCmd.Flags().StringVar(&o.ServiceRepoURL, "service-repo-url", "", "Provide the URL for your Service repository e.g. https://github.com/organisation/service.git")
 	bootstrapCmd.Flags().StringVar(&o.ServiceWebhookSecret, "service-webhook-secret", "", "Provide a secret that we can use to authenticate incoming hooks from your Git hosting service for the Service repository. (if not provided, it will be auto-generated)")
 	bootstrapCmd.Flags().StringVar(&o.PrivateRepoDriver, "private-repo-driver", "", "If your Git repositories are on a custom domain, please indicate which driver to use github or gitlab")
+	bootstrapCmd.Flags().BoolVar(&o.CommitStatusTracker, "commit-status-tracker", true, "Enable or disable the commit-status-tracker which reports the success/failure of your pipelineruns to GitHub/GitLab")
 	return bootstrapCmd
 }
 
@@ -307,10 +310,7 @@ func isKnownDriver(repoURL string) bool {
 		return false
 	}
 	_, err = factory.DefaultIdentifier.Identify(host)
-	if err == nil {
-		return true
-	}
-	return false
+	return err == nil
 }
 
 func hostFromURL(s string) (string, error) {

--- a/pkg/odo/cli/pipelines/bootstrap_test.go
+++ b/pkg/odo/cli/pipelines/bootstrap_test.go
@@ -9,6 +9,7 @@ import (
 
 	"k8s.io/client-go/kubernetes"
 
+	"github.com/google/go-cmp/cmp"
 	"github.com/openshift/odo/pkg/odo/genericclioptions"
 	"github.com/openshift/odo/pkg/pipelines"
 	appv1 "k8s.io/api/apps/v1"
@@ -23,10 +24,6 @@ type mockSpinner struct {
 	writer io.Writer
 	start  bool
 	end    bool
-}
-type keyValuePair struct {
-	key   string
-	value string
 }
 
 func TestValidatePrefix(t *testing.T) {
@@ -209,7 +206,7 @@ func TestDependenciesWithNothingInstalled(t *testing.T) {
 
 	wantMsg := `
 Checking if Sealed Secrets is installed with the default configuration[Please install Sealed Secrets from https://github.com/bitnami-labs/sealed-secrets/releases]
-Checking if ArgoCD Operator is installed with the default configuration[Please install ArgoCD operator from OperatorHub]
+Checking if ArgoCD Operator is installed with the default configuration[Please install ArgoCD operator from OperatorHub, with an ArgoCD resource called 'argocd']
 Checking if OpenShift Pipelines Operator is installed with the default configuration[Please install OpenShift Pipelines operator from OperatorHub]`
 
 	buff := &bytes.Buffer{}
@@ -246,7 +243,7 @@ func TestDependenciesWithNoArgoCD(t *testing.T) {
 
 	wantMsg := `
 Checking if Sealed Secrets is installed with the default configuration
-Checking if ArgoCD Operator is installed with the default configuration[Please install ArgoCD operator from OperatorHub]
+Checking if ArgoCD Operator is installed with the default configuration[Please install ArgoCD operator from OperatorHub, with an ArgoCD resource called 'argocd']
 Checking if OpenShift Pipelines Operator is installed with the default configuration`
 
 	buff := &bytes.Buffer{}
@@ -281,19 +278,19 @@ func assertError(t *testing.T, err error, msg string) {
 	t.Helper()
 	if err == nil {
 		if msg != "" {
-			t.Fatalf("Error mismatch: got %v, want %v", err, msg)
+			t.Fatalf("error mismatch: got %v, want %v", err, msg)
 		}
 		return
 	}
 	if err.Error() != msg {
-		t.Fatalf("Error mismatch: got %s, want %s", err.Error(), msg)
+		t.Fatalf("error mismatch: got %s, want %s", err.Error(), msg)
 	}
 }
 
 func assertMessage(t *testing.T, got, want string) {
 	t.Helper()
-	if got != want {
-		t.Fatalf("Message mismatch: got %s, want %s", got, want)
+	if diff := cmp.Diff(want, got); diff != "" {
+		t.Fatalf("message comparison failed:\n%s", diff)
 	}
 }
 
@@ -345,13 +342,6 @@ func matchError(t *testing.T, s string, e error) bool {
 		t.Fatal(err)
 	}
 	return match
-}
-
-func flag(k, v string) keyValuePair {
-	return keyValuePair{
-		key:   k,
-		value: v,
-	}
 }
 
 func (m *mockSpinner) Start(status string, debug bool) {

--- a/pkg/odo/cli/pipelines/environment/add.go
+++ b/pkg/odo/cli/pipelines/environment/add.go
@@ -30,7 +30,6 @@ var (
 // AddEnvParameters encapsulates the parameters for the odo pipelines init command.
 type AddEnvParameters struct {
 	envName         string
-	output          string
 	pipelinesFolder string
 	cluster         string
 	// generic context options common to all commands

--- a/pkg/odo/cli/pipelines/environment/add_test.go
+++ b/pkg/odo/cli/pipelines/environment/add_test.go
@@ -2,7 +2,6 @@ package environment
 
 import (
 	"bytes"
-	"regexp"
 	"testing"
 
 	"github.com/spf13/cobra"
@@ -40,21 +39,6 @@ func executeCommand(cmd *cobra.Command, flags ...keyValuePair) (c *cobra.Command
 	}
 	c, err = cmd.ExecuteC()
 	return c, buf.String(), err
-}
-
-func matchError(t *testing.T, s string, e error) bool {
-	t.Helper()
-	if s == "" && e == nil {
-		return true
-	}
-	if s != "" && e == nil {
-		return false
-	}
-	match, err := regexp.MatchString(s, e.Error())
-	if err != nil {
-		t.Fatal(err)
-	}
-	return match
 }
 
 func flag(k, v string) keyValuePair {

--- a/pkg/odo/cli/pipelines/ui/ui.go
+++ b/pkg/odo/cli/pipelines/ui/ui.go
@@ -89,7 +89,7 @@ func EnterOutputPath() string {
 	var outputPath string
 	prompt := &survey.Input{
 		Message: "Provide a path to write GitOps resources?",
-		Help:    fmt.Sprintf("This is the path where the GitOps repository configuration is stored locally before you push it to the repository GitopsRepoURL"),
+		Help:    "This is the path where the GitOps repository configuration is stored locally before you push it to the repository GitopsRepoURL",
 		Default: ".",
 	}
 
@@ -108,8 +108,7 @@ func EnterOutputPath() string {
 // EnterGitWebhookSecret allows the user to specify the webhook secret string they wish to authenticate push/pull to GitOps repo in a UI prompt.
 func EnterGitWebhookSecret() string {
 	var gitWebhookSecret string
-	var prompt *survey.Input
-	prompt = &survey.Input{
+	prompt := &survey.Input{
 		Message: "Provide a secret (minimum 16 characters) that we can use to authenticate incoming hooks from your Git hosting service for the Service repository. (if not provided, it will be auto-generated)",
 		Help:    "You can provide a string that is used as a shared secret to authenticate the origin of hook notifications from your git host.",
 	}
@@ -144,11 +143,12 @@ func EnterSealedSecretNamespace() string {
 	return sealedNs
 }
 
-// EnterStatusTrackerAccessToken , it becomes necessary to add the personal access token from github to autheticate the commit-status-tracker.
-func EnterStatusTrackerAccessToken(serviceRepo string) string {
+// EnterGitHostAccessToken , it becomes necessary to add the personal access
+// token to access upstream git hosts.
+func EnterGitHostAccessToken(serviceRepo string) string {
 	var accessToken string
 	prompt := &survey.Password{
-		Message: "Please provide a token used to authenticate API calls to push commit-status updates to your Git hosting service",
+		Message: fmt.Sprintf("Please provide a token used to authenticate requests to %q", serviceRepo),
 		Help:    "commit-status-tracker reports the completion status of OpenShift pipeline runs to your Git hosting status on success or failure, this token will be encrypted as a secret in your cluster.\nIf you are using Github, please see here for how to generate a token https://docs.github.com/en/github/authenticating-to-github/creating-a-personal-access-token\nIf you are using GitLab, please see here for how to generate a token https://docs.gitlab.com/ee/user/profile/personal_access_tokens.html",
 	}
 	err := survey.AskOne(prompt, &accessToken, makeAccessTokenCheck(serviceRepo))
@@ -232,9 +232,9 @@ func SelectOptionCommitStatusTracker() string {
 	var optionCommitStatusTracker string
 	prompt := &survey.Select{
 		Message: "Do you want to enable commit-status-tracker?",
+		Help:    "commit-status-tracker reports the completion status of OpenShift pipeline runs to your git host on success or failure",
 		Options: []string{"yes", "no"},
 	}
-
 	err := survey.AskOne(prompt, &optionCommitStatusTracker, survey.Required)
 	ui.HandleError(err)
 	return optionCommitStatusTracker
@@ -252,4 +252,17 @@ func SelectPrivateRepoDriver() string {
 	err := survey.AskOne(prompt, &driver, survey.Required)
 	ui.HandleError(err)
 	return driver
+}
+
+// IsPrivateRepo lets the user choose between a private / public repository.
+func IsPrivateRepo() bool {
+	var response string
+	prompt := &survey.Select{
+		Message: "Is this repository a private repository?",
+		Options: []string{"yes", "no"},
+	}
+
+	err := survey.AskOne(prompt, &response, survey.Required)
+	ui.HandleError(err)
+	return response == "yes"
 }

--- a/pkg/odo/cli/pipelines/ui/validate.go
+++ b/pkg/odo/cli/pipelines/ui/validate.go
@@ -95,7 +95,13 @@ func validateAccessToken(input interface{}, serviceRepo string) error {
 	if s, ok := input.(string); ok {
 		repo, _ := git.NewRepository(serviceRepo, s)
 		parsedURL, err := url.Parse(serviceRepo)
+		if err != nil {
+			return fmt.Errorf("failed to parse the provided URL %q: %w", serviceRepo, err)
+		}
 		repoName, err := git.GetRepoName(parsedURL)
+		if err != nil {
+			return fmt.Errorf("failed to get the repository name from %q: %w", serviceRepo, err)
+		}
 		_, _, err = repo.Client.Repositories.Find(context.Background(), repoName)
 		if err != nil {
 			return fmt.Errorf("The token passed is incorrect for repository %s", repoName)

--- a/pkg/pipelines/bootstrap.go
+++ b/pkg/pipelines/bootstrap.go
@@ -4,7 +4,6 @@ import (
 	"errors"
 	"fmt"
 	"net/url"
-	"os"
 	"path/filepath"
 	"sort"
 	"strings"
@@ -46,6 +45,8 @@ const (
 	rolebindingsPath      = "02-rolebindings/pipeline-service-rolebinding.yaml"
 	serviceAccountPath    = "02-rolebindings/pipeline-service-account.yaml"
 	secretsPath           = "03-secrets/gitops-webhook-secret.yaml"
+	authTokenPath         = "03-secrets/git-host-access-token.yaml"
+	basicAuthTokenPath    = "03-secrets/git-host-basic-auth-token.yaml"
 	dockerConfigPath      = "03-secrets/docker-config.yaml"
 	gitopsTasksPath       = "04-tasks/deploy-from-source-task.yaml"
 	appTaskPath           = "04-tasks/deploy-using-kubectl-task.yaml"
@@ -79,11 +80,12 @@ type BootstrapOptions struct {
 	InternalRegistryHostname string               // This is the internal registry hostname used for pushing images.
 	OutputPath               string               // Where to write the bootstrapped files to?
 	SealedSecretsService     types.NamespacedName // SealedSecrets Services name
-	StatusTrackerAccessToken string               // The auth token to use to send commit-status notifications.
+	GitHostAccessToken       string               // The auth token to use to send commit-status notifications, and access private repositories.
 	Overwrite                bool                 // This allows to overwrite if there is an exixting gitops repository
 	ServiceRepoURL           string               // This is the full URL to your GitHub repository for your app source.
 	ServiceWebhookSecret     string               // This is the secret for authenticating hooks from your app source.
 	PrivateRepoDriver        string               // Records the type of the GitOpsRepoURL driver if not a well-known host.
+	CommitStatusTracker      bool                 // If true, this is a "private repository", i.e. requires authentication to clone the repository.
 }
 
 // PolicyRules to be bound to service account
@@ -448,7 +450,7 @@ func createDockerSecret(fs afero.Fs, dockerConfigJSONFilename, secretNS string, 
 	return dockerSecret, nil
 }
 
-// createCICDResources creates resources assocated to pipelines.
+// createCICDResources creates resources for OpenShift pipelines.
 func createCICDResources(fs afero.Fs, repo scm.Repository, pipelineConfig *config.PipelinesConfig, o *BootstrapOptions) (res.Resources, error) {
 	cicdNamespace := pipelineConfig.Name
 	// key: path of the resource
@@ -475,10 +477,15 @@ func createCICDResources(fs afero.Fs, repo scm.Repository, pipelineConfig *confi
 		outputs[serviceAccountPath] = roles.AddSecretToSA(sa, dockerSecretName)
 	}
 
-	if o.StatusTrackerAccessToken != "" {
-		trackerResources, err := statustracker.Resources(cicdNamespace, o.StatusTrackerAccessToken, o.SealedSecretsService,
-			o.GitOpsRepoURL, o.PrivateRepoDriver)
+	if o.GitHostAccessToken != "" {
+		err := generateSecrets(outputs, sa, cicdNamespace, o)
+		if err != nil {
+			return nil, err
+		}
+	}
 
+	if o.CommitStatusTracker {
+		trackerResources, err := statustracker.Resources(cicdNamespace, o.GitOpsRepoURL, o.PrivateRepoDriver)
 		if err != nil {
 			return nil, err
 		}
@@ -548,21 +555,28 @@ func getResourceFiles(res res.Resources) []string {
 	return files
 }
 
-// check if the file exists or not
-func CheckFileExists(fs afero.Fs, dockerConfigJSONFilename string) (string, error) {
-	if dockerConfigJSONFilename == "" {
-		return "", errors.New("failed to generate path to file: must provide --dockerconfigjson parameter")
+func generateSecrets(outputs res.Resources, sa *corev1.ServiceAccount, ns string, o *BootstrapOptions) error {
+	if o.CommitStatusTracker {
+		tokenSecret, err := secrets.CreateSealedSecret(meta.NamespacedName(
+			ns, "git-host-access-token"), o.SealedSecretsService, o.GitHostAccessToken, "token")
+		if err != nil {
+			return fmt.Errorf("failed to generate access token Secret: %w", err)
+		}
+		outputs[authTokenPath] = tokenSecret
+		outputs[serviceAccountPath] = roles.AddSecretToSA(sa, tokenSecret.Name)
 	}
-	authJSONPath, err := homedir.Expand(dockerConfigJSONFilename)
+	secretTargetHost, err := repoURL(o.ServiceRepoURL)
 	if err != nil {
-		return "", fmt.Errorf("failed to generate path to file: %v", err)
+		return fmt.Errorf("failed to parse the Service Repo URL %q: %w", o.ServiceRepoURL, err)
 	}
-	_, err = fs.Stat(authJSONPath)
-	if os.IsNotExist(err) {
-		return "", fmt.Errorf("file does not exist at the file-path passed to access the dockercfgjson, kindly enter a valid file path")
-	}
+	basicAuthSecret, err := secrets.CreateSealedBasicAuthSecret(meta.NamespacedName(
+		ns, "git-host-basic-auth-token"), o.SealedSecretsService, o.GitHostAccessToken, meta.AddAnnotations(map[string]string{
+		"tekton.dev/git-0": secretTargetHost,
+	}))
 	if err != nil {
-		return "", fmt.Errorf("failed to read Docker config %q : %s", authJSONPath, err)
+		return fmt.Errorf("failed to generate basic auth token Secret: %w", err)
 	}
-	return authJSONPath, nil
+	outputs[basicAuthTokenPath] = basicAuthSecret
+	outputs[serviceAccountPath] = roles.AddSecretToSA(sa, basicAuthSecret.Name)
+	return nil
 }

--- a/pkg/pipelines/bootstrap_test.go
+++ b/pkg/pipelines/bootstrap_test.go
@@ -5,17 +5,21 @@ import (
 	"crypto/rsa"
 	"testing"
 
+	ssv1alpha1 "github.com/bitnami-labs/sealed-secrets/pkg/apis/sealed-secrets/v1alpha1"
 	"github.com/google/go-cmp/cmp"
 	"github.com/google/go-cmp/cmp/cmpopts"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+
 	"github.com/openshift/odo/pkg/pipelines/config"
 	"github.com/openshift/odo/pkg/pipelines/deployment"
 	"github.com/openshift/odo/pkg/pipelines/eventlisteners"
 	"github.com/openshift/odo/pkg/pipelines/ioutils"
 	"github.com/openshift/odo/pkg/pipelines/meta"
 	res "github.com/openshift/odo/pkg/pipelines/resources"
+	"github.com/openshift/odo/pkg/pipelines/roles"
 	"github.com/openshift/odo/pkg/pipelines/scm"
 	"github.com/openshift/odo/pkg/pipelines/secrets"
-	"k8s.io/apimachinery/pkg/types"
 )
 
 const (
@@ -41,20 +45,19 @@ func TestBootstrapManifest(t *testing.T) {
 	}
 
 	params := &BootstrapOptions{
-		Prefix:                   "tst-",
-		GitOpsRepoURL:            testGitOpsRepo,
-		ImageRepo:                "image/repo",
-		GitOpsWebhookSecret:      "123",
-		StatusTrackerAccessToken: "test-token",
-
+		Prefix:               "tst-",
+		GitOpsRepoURL:        testGitOpsRepo,
+		ImageRepo:            "image/repo",
+		GitOpsWebhookSecret:  "123",
+		GitHostAccessToken:   "test-token",
 		ServiceRepoURL:       testSvcRepo,
 		ServiceWebhookSecret: "456",
+		CommitStatusTracker:  true,
 	}
 
 	r, err := bootstrapResources(params, ioutils.NewMemoryFilesystem())
-	if err != nil {
-		t.Fatal(err)
-	}
+	fatalIfError(t, err)
+
 	hookSecret, err := secrets.CreateSealedSecret(
 		meta.NamespacedName("tst-cicd", "webhook-secret-tst-dev-http-api"),
 		meta.NamespacedName("test-ns", "service"), "456", eventlisteners.WebhookSecretKey)
@@ -127,9 +130,11 @@ func TestBootstrapManifest(t *testing.T) {
 		"02-rolebindings/commit-status-tracker-rolebinding.yaml",
 		"02-rolebindings/commit-status-tracker-service-account.yaml",
 		"02-rolebindings/internal-registry-image-binding.yaml",
+		"02-rolebindings/pipeline-service-account.yaml",
 		"02-rolebindings/pipeline-service-role.yaml",
 		"02-rolebindings/pipeline-service-rolebinding.yaml",
-		"03-secrets/commit-status-tracker.yaml",
+		"03-secrets/git-host-access-token.yaml",
+		"03-secrets/git-host-basic-auth-token.yaml",
 		"03-secrets/gitops-webhook-secret.yaml",
 		"03-secrets/webhook-secret-tst-dev-http-api.yaml",
 		"04-tasks/deploy-from-source-task.yaml",
@@ -146,16 +151,14 @@ func TestBootstrapManifest(t *testing.T) {
 	}
 	k := r["config/tst-cicd/base/kustomization.yaml"].(res.Kustomization)
 	if diff := cmp.Diff(wantResources, k.Resources); diff != "" {
-		t.Fatalf("did not add the secret to the base kustomization: %s\n", diff)
+		t.Fatalf("base kustomization failed:\n%s\n", diff)
 	}
 }
 
 func TestOrgRepoFromURL(t *testing.T) {
 	want := "my-org/gitops"
 	got, err := orgRepoFromURL(testGitOpsRepo)
-	if err != nil {
-		t.Fatal(err)
-	}
+	fatalIfError(t, err)
 	if got != want {
 		t.Fatalf("orgRepFromURL(%s) got %s, want %s", testGitOpsRepo, got, want)
 	}
@@ -176,9 +179,8 @@ func TestApplicationFromRepo(t *testing.T) {
 	}
 
 	got, err := applicationFromRepo(testSvcRepo, svc)
-	if err != nil {
-		t.Fatal(err)
-	}
+	fatalIfError(t, err)
+
 	if diff := cmp.Diff(want, got); diff != "" {
 		t.Fatalf("bootstrapped resources:\n%s", diff)
 	}
@@ -206,9 +208,8 @@ func TestOverwriteFlag(t *testing.T) {
 		ServiceWebhookSecret: "456",
 	}
 	err := Bootstrap(params, fakeFs)
-	if err != nil {
-		t.Fatal(err)
-	}
+	fatalIfError(t, err)
+
 	got := Bootstrap(params, fakeFs)
 	want := "pipelines.yaml in output path already exists. If you want to replace your existing files, please rerun with --overwrite."
 	if diff := cmp.Diff(want, got.Error()); diff != "" {
@@ -277,7 +278,6 @@ func TestGetCICDKustomization(t *testing.T) {
 	if diff := cmp.Diff(want, got); diff != "" {
 		t.Fatalf("getCICDKustomization was not correct: %s\n", diff)
 	}
-
 }
 
 func TestAddPrefixToResources(t *testing.T) {
@@ -300,5 +300,133 @@ func TestAddPrefixToResources(t *testing.T) {
 	}
 	if diff := cmp.Diff(want, addPrefixToResources("test-prefix", files)); diff != "" {
 		t.Fatalf("addPrefixToResources failed, diff %s\n", diff)
+	}
+}
+
+func TestGenerateSecrets(t *testing.T) {
+	defer stubDefaultPublicKeyFunc(t)()
+	ns := "test-ns"
+	outputs := res.Resources{}
+	sa := roles.CreateServiceAccount(meta.NamespacedName("test-ns", "test-sa"))
+	o := &BootstrapOptions{
+		SealedSecretsService: meta.NamespacedName("sealed-secrets", "secrets"),
+		GitHostAccessToken:   "abc123",
+		ServiceRepoURL:       "https://gl.example.com/my-org/my-project.git",
+		CommitStatusTracker:  true,
+	}
+
+	err := generateSecrets(outputs, sa, ns, o)
+	fatalIfError(t, err)
+
+	wantSA := &corev1.ServiceAccount{
+		TypeMeta: meta.TypeMeta("ServiceAccount", "v1"),
+		ObjectMeta: meta.ObjectMeta(
+			types.NamespacedName{Name: "test-sa", Namespace: "test-ns"},
+		),
+		Secrets: []corev1.ObjectReference{{Name: "git-host-access-token"}, {Name: "git-host-basic-auth-token"}},
+	}
+	if diff := cmp.Diff(wantSA, outputs[serviceAccountPath]); diff != "" {
+		t.Fatalf("generatedSecrets failed to update the ServiceAccount:\n%s", diff)
+	}
+
+	wantBasicAuthSecret := &ssv1alpha1.SealedSecret{
+		TypeMeta: meta.TypeMeta("SealedSecret", "bitnami.com/v1alpha1"),
+		ObjectMeta: meta.ObjectMeta(
+			types.NamespacedName{Name: "git-host-basic-auth-token", Namespace: "test-ns"},
+		),
+		Spec: ssv1alpha1.SealedSecretSpec{
+			Template: ssv1alpha1.SecretTemplateSpec{
+				ObjectMeta: meta.ObjectMeta(
+					types.NamespacedName{Name: "git-host-basic-auth-token", Namespace: "test-ns"},
+					meta.AddAnnotations(
+						map[string]string{
+							"tekton.dev/git-0": "https://gl.example.com",
+						}),
+				),
+				Type: corev1.SecretTypeBasicAuth,
+			},
+		},
+	}
+	if diff := cmp.Diff(wantBasicAuthSecret, outputs[basicAuthTokenPath],
+		cmpopts.IgnoreFields(ssv1alpha1.SealedSecret{}, "Spec.EncryptedData", "ObjectMeta.Annotations")); diff != "" {
+		t.Fatalf("generatedSecrets failed to create basic auth token secret:\n%s", diff)
+	}
+	wantAuthSecret := &ssv1alpha1.SealedSecret{
+		TypeMeta: meta.TypeMeta("SealedSecret", "bitnami.com/v1alpha1"),
+		ObjectMeta: meta.ObjectMeta(
+			types.NamespacedName{Name: "git-host-access-token", Namespace: "test-ns"},
+		),
+		Spec: ssv1alpha1.SealedSecretSpec{
+			Template: ssv1alpha1.SecretTemplateSpec{
+				ObjectMeta: meta.ObjectMeta(
+					types.NamespacedName{Name: "git-host-access-token", Namespace: "test-ns"},
+				),
+				Type: corev1.SecretTypeOpaque,
+			},
+		},
+	}
+	if diff := cmp.Diff(wantAuthSecret, outputs[authTokenPath],
+		cmpopts.IgnoreFields(ssv1alpha1.SealedSecret{}, "Spec.EncryptedData", "ObjectMeta.Annotations")); diff != "" {
+		t.Fatalf("generatedSecrets failed to create auth token secret:\n%s", diff)
+	}
+}
+
+func TestGenerateSecretsWithNoCommitStatusTracker(t *testing.T) {
+	defer stubDefaultPublicKeyFunc(t)()
+	ns := "test-ns"
+	outputs := res.Resources{}
+	sa := roles.CreateServiceAccount(meta.NamespacedName("test-ns", "test-sa"))
+	o := &BootstrapOptions{
+		SealedSecretsService: meta.NamespacedName("sealed-secrets", "secrets"),
+		GitHostAccessToken:   "abc123",
+		ServiceRepoURL:       "https://gl.example.com/my-org/my-project.git",
+		CommitStatusTracker:  false,
+	}
+
+	err := generateSecrets(outputs, sa, ns, o)
+	fatalIfError(t, err)
+
+	wantSA := &corev1.ServiceAccount{
+		TypeMeta: meta.TypeMeta("ServiceAccount", "v1"),
+		ObjectMeta: meta.ObjectMeta(
+			types.NamespacedName{Name: "test-sa", Namespace: "test-ns"},
+		),
+		Secrets: []corev1.ObjectReference{{Name: "git-host-basic-auth-token"}},
+	}
+	if diff := cmp.Diff(wantSA, outputs[serviceAccountPath]); diff != "" {
+		t.Fatalf("generatedSecrets failed to update the ServiceAccount:\n%s", diff)
+	}
+
+	wantBasicAuthSecret := &ssv1alpha1.SealedSecret{
+		TypeMeta: meta.TypeMeta("SealedSecret", "bitnami.com/v1alpha1"),
+		ObjectMeta: meta.ObjectMeta(
+			types.NamespacedName{Name: "git-host-basic-auth-token", Namespace: "test-ns"},
+		),
+		Spec: ssv1alpha1.SealedSecretSpec{
+			Template: ssv1alpha1.SecretTemplateSpec{
+				ObjectMeta: meta.ObjectMeta(
+					types.NamespacedName{Name: "git-host-basic-auth-token", Namespace: "test-ns"},
+					meta.AddAnnotations(
+						map[string]string{
+							"tekton.dev/git-0": "https://gl.example.com",
+						}),
+				),
+				Type: corev1.SecretTypeBasicAuth,
+			},
+		},
+	}
+	if diff := cmp.Diff(wantBasicAuthSecret, outputs[basicAuthTokenPath],
+		cmpopts.IgnoreFields(ssv1alpha1.SealedSecret{}, "Spec.EncryptedData", "ObjectMeta.Annotations")); diff != "" {
+		t.Fatalf("generatedSecrets failed to create basic auth token secret:\n%s", diff)
+	}
+	if outputs[authTokenPath] != nil {
+		t.Fatalf("auth token secret for commit status tracker generated: %#v", outputs[authTokenPath])
+	}
+}
+
+func fatalIfError(t *testing.T, err error) {
+	t.Helper()
+	if err != nil {
+		t.Fatal(err)
 	}
 }

--- a/pkg/pipelines/meta/meta.go
+++ b/pkg/pipelines/meta/meta.go
@@ -14,7 +14,7 @@ func TypeMeta(kind, apiVersion string) metav1.TypeMeta {
 }
 
 // ObjectMeta creates metav1.ObjectMeta
-func ObjectMeta(n types.NamespacedName, opts ...objectMetaFunc) metav1.ObjectMeta {
+func ObjectMeta(n types.NamespacedName, opts ...ObjectMetaOpt) metav1.ObjectMeta {
 	om := metav1.ObjectMeta{
 		Namespace: n.Namespace,
 		Name:      n.Name,
@@ -27,7 +27,7 @@ func ObjectMeta(n types.NamespacedName, opts ...objectMetaFunc) metav1.ObjectMet
 
 // AddLabels is an option func for the ObjectMeta function, which additively
 // applies the provided labels to the created ObjectMeta.
-func AddLabels(l map[string]string) objectMetaFunc {
+func AddLabels(l map[string]string) ObjectMetaOpt {
 	return func(om *metav1.ObjectMeta) {
 		if om.Labels == nil {
 			om.Labels = map[string]string{}
@@ -38,7 +38,22 @@ func AddLabels(l map[string]string) objectMetaFunc {
 	}
 }
 
-type objectMetaFunc func(om *metav1.ObjectMeta)
+// AddAnnotations is an option func for the ObjectMeta function, which additively
+// applies the provided labels to the created ObjectMeta.
+func AddAnnotations(l map[string]string) ObjectMetaOpt {
+	return func(om *metav1.ObjectMeta) {
+		if om.Annotations == nil {
+			om.Annotations = map[string]string{}
+		}
+		for k, v := range l {
+			om.Annotations[k] = v
+		}
+	}
+}
+
+// ObjectMetaOpt is a function that can change a newly created meta.ObjectMeta
+// when it's being created.
+type ObjectMetaOpt func(om *metav1.ObjectMeta)
 
 // NamespacedName creates types.NamespacedName
 func NamespacedName(ns, name string) types.NamespacedName {

--- a/pkg/pipelines/meta/meta_test.go
+++ b/pkg/pipelines/meta/meta_test.go
@@ -12,13 +12,43 @@ func TestLabels(t *testing.T) {
 		"app":     "my-app",
 		"service": "my-service",
 	})
-	om := &metav1.ObjectMeta{}
+	om := &metav1.ObjectMeta{
+		Labels: map[string]string{
+			"my-label": "my-value",
+		},
+	}
 	v(om)
 
 	want := &metav1.ObjectMeta{
 		Labels: map[string]string{
-			"app":     "my-app",
-			"service": "my-service",
+			"my-label": "my-value",
+			"app":      "my-app",
+			"service":  "my-service",
+		},
+	}
+
+	if diff := cmp.Diff(want, om); diff != "" {
+		t.Fatalf("failed to add labels:\n%s", diff)
+	}
+}
+
+func TestAnnotations(t *testing.T) {
+	v := AddAnnotations(map[string]string{
+		"app":     "my-app",
+		"service": "my-service",
+	})
+	om := &metav1.ObjectMeta{
+		Annotations: map[string]string{
+			"my-annotation": "my-value",
+		},
+	}
+	v(om)
+
+	want := &metav1.ObjectMeta{
+		Annotations: map[string]string{
+			"my-annotation": "my-value",
+			"app":           "my-app",
+			"service":       "my-service",
 		},
 	}
 

--- a/pkg/pipelines/namespaces/namespaces.go
+++ b/pkg/pipelines/namespaces/namespaces.go
@@ -61,11 +61,11 @@ func Create(name, gitOpsRepoURL string) *corev1.Namespace {
 func GetClientSet() (*kubernetes.Clientset, error) {
 	clientConfig, err := clientconfig.GetRESTConfig()
 	if err != nil {
-		return nil, fmt.Errorf("failed to get client config due to %v", err)
+		return nil, fmt.Errorf("failed to get Kubernetes client config: %w", err)
 	}
 	clientSet, err := kubernetes.NewForConfig(clientConfig)
 	if err != nil {
-		return nil, fmt.Errorf("failed to get APIs client due to %v", err)
+		return nil, fmt.Errorf("failed to create Kubernetes API client: %w", err)
 	}
 	return clientSet, nil
 }

--- a/pkg/pipelines/secrets/secrets.go
+++ b/pkg/pipelines/secrets/secrets.go
@@ -56,6 +56,12 @@ func CreateSealedSecret(name, service types.NamespacedName, data, secretKey stri
 	return seal(secret, DefaultPublicKeyFunc, service)
 }
 
+// CreateSealedBasicAuthSecret creates a SealedSecret with a BasicAuth type
+// secret.
+func CreateSealedBasicAuthSecret(name, service types.NamespacedName, token string, opts ...meta.ObjectMetaOpt) (*ssv1alpha1.SealedSecret, error) {
+	return seal(createBasicAuthSecret(name, token, opts...), DefaultPublicKeyFunc, service)
+}
+
 // Returns a sealed secret
 func seal(secret *corev1.Secret, pubKey PublicKeyFunc, service types.NamespacedName) (*ssv1alpha1.SealedSecret, error) {
 	// Strip read-only server-side ObjectMeta (if present)
@@ -139,9 +145,8 @@ func parseKey(r io.Reader) (*rsa.PublicKey, error) {
 func getRESTClient() (*clientv1.CoreV1Client, error) {
 	config, err := clientconfig.GetRESTConfig()
 	if err != nil {
-		return nil, fmt.Errorf("failed to get client config due to %v", err)
+		return nil, fmt.Errorf("failed to get Kubernetes client config: %w", err)
 	}
-
 	config.AcceptContentTypes = "application/x-pem-file, */*"
 	return clientv1.NewForConfig(config)
 }
@@ -157,6 +162,18 @@ func createOpaqueSecret(name types.NamespacedName, data, secretKey string) (*cor
 // body, and type DockerConfigJson.
 func createDockerConfigSecret(name types.NamespacedName, in io.Reader) (*corev1.Secret, error) {
 	return createSecret(name, ".dockerconfigjson", corev1.SecretTypeDockerConfigJson, in)
+}
+
+func createBasicAuthSecret(name types.NamespacedName, token string, opts ...meta.ObjectMetaOpt) *corev1.Secret {
+	return &corev1.Secret{
+		TypeMeta:   secretTypeMeta,
+		ObjectMeta: meta.ObjectMeta(name, opts...),
+		Type:       corev1.SecretTypeBasicAuth,
+		StringData: map[string]string{
+			"username": "tekton",
+			"password": token,
+		},
+	}
 }
 
 func createSecret(name types.NamespacedName, key string, st corev1.SecretType, in io.Reader) (*corev1.Secret, error) {

--- a/pkg/pipelines/secrets/secrets_test.go
+++ b/pkg/pipelines/secrets/secrets_test.go
@@ -235,7 +235,6 @@ func TestSeal(t *testing.T) {
 					}
 				}
 			}
-
 			// NB: See sealedsecret_test.go for e2e crypto test
 		})
 	}
@@ -301,6 +300,36 @@ func TestCreateDockerConfigSecret(t *testing.T) {
 
 	if diff := cmp.Diff(want, secret); diff != "" {
 		t.Fatalf("createDockerConfigSecret() failed got\n%s", diff)
+	}
+}
+
+func TestBasicAuthSecret(t *testing.T) {
+	token := "abcdefghijklmnop"
+	host := "https://github.com"
+	secret := createBasicAuthSecret(meta.NamespacedName("cicd", "github-auth"), token, meta.AddAnnotations(
+		map[string]string{
+			"tekton.dev/git-0": host,
+		}),
+	)
+
+	want := &corev1.Secret{
+		TypeMeta: secretTypeMeta,
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "github-auth",
+			Namespace: "cicd",
+			Annotations: map[string]string{
+				"tekton.dev/git-0": host,
+			},
+		},
+		Type: corev1.SecretTypeBasicAuth,
+		StringData: map[string]string{
+			"username": "tekton",
+			"password": token,
+		},
+	}
+
+	if diff := cmp.Diff(want, secret); diff != "" {
+		t.Fatalf("createBasicAuthSecret() failed got\n%s", diff)
 	}
 }
 

--- a/pkg/pipelines/statustracker/deployment.go
+++ b/pkg/pipelines/statustracker/deployment.go
@@ -68,30 +68,25 @@ var (
 	}
 )
 
+// Resources returns a list of newly created resources that are required to
+// setup the status-tracker service.
+func Resources(ns, repoURL, driver string) (res.Resources, error) {
+	name := meta.NamespacedName(ns, operatorName)
+	sa := roles.CreateServiceAccount(name)
+
+	return res.Resources{
+		"02-rolebindings/commit-status-tracker-role.yaml":            roles.CreateRole(name, roleRules),
+		"02-rolebindings/commit-status-tracker-rolebinding.yaml":     roles.CreateRoleBinding(name, sa, "Role", operatorName),
+		"02-rolebindings/commit-status-tracker-service-account.yaml": sa,
+		"10-commit-status-tracker/operator.yaml":                     createStatusTrackerDeployment(ns, repoURL, driver),
+	}, nil
+}
+
 func createStatusTrackerDeployment(ns, repoURL, driver string) *appsv1.Deployment {
 	return deployment.Create(commitStatusAppLabel, ns, operatorName, containerImage,
 		deployment.ServiceAccount(operatorName),
 		deployment.Env(makeEnvironment(repoURL, driver)),
 		deployment.Command([]string{operatorName}))
-}
-
-// Resources returns a list of newly created resources that are required start
-// the status-tracker service.
-func Resources(ns, token string, sealedSecretsservice types.NamespacedName, repoURL, driver string) (res.Resources, error) {
-	name := meta.NamespacedName(ns, operatorName)
-	sa := roles.CreateServiceAccount(name)
-
-	githubAuth, err := defaultSecretSealer(meta.NamespacedName(ns, "commit-status-tracker-git-secret"), sealedSecretsservice, token, "token")
-	if err != nil {
-		return nil, fmt.Errorf("failed to generate Status Tracker Secret: %v", err)
-	}
-	return res.Resources{
-		"02-rolebindings/commit-status-tracker-role.yaml":            roles.CreateRole(name, roleRules),
-		"02-rolebindings/commit-status-tracker-rolebinding.yaml":     roles.CreateRoleBinding(name, sa, "Role", operatorName),
-		"02-rolebindings/commit-status-tracker-service-account.yaml": sa,
-		"03-secrets/commit-status-tracker.yaml":                      githubAuth,
-		"10-commit-status-tracker/operator.yaml":                     createStatusTrackerDeployment(ns, repoURL, driver),
-	}, nil
 }
 
 func ptr32(i int32) *int32 {

--- a/pkg/pipelines/statustracker/deployment_test.go
+++ b/pkg/pipelines/statustracker/deployment_test.go
@@ -6,9 +6,7 @@ import (
 	appsv1 "k8s.io/api/apps/v1"
 	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	"k8s.io/apimachinery/pkg/types"
 
-	ssv1alpha1 "github.com/bitnami-labs/sealed-secrets/pkg/apis/sealed-secrets/v1alpha1"
 	"github.com/google/go-cmp/cmp"
 	"github.com/openshift/odo/pkg/pipelines/deployment"
 	"github.com/openshift/odo/pkg/pipelines/meta"
@@ -81,17 +79,8 @@ func TestCreateStatusTrackerDeployment(t *testing.T) {
 }
 
 func TestResource(t *testing.T) {
-	defer func(f secretSealer) {
-		defaultSecretSealer = f
-	}(defaultSecretSealer)
-
-	testSecret := &ssv1alpha1.SealedSecret{}
-	defaultSecretSealer = func(ns, _ types.NamespacedName, data, secretKey string) (*ssv1alpha1.SealedSecret, error) {
-		return testSecret, nil
-	}
-
 	ns := "my-test-ns"
-	generated, err := Resources(ns, "test-token", meta.NamespacedName("sealed-secrets-ns", "sealed-secrets-svc"), testRepoURL, "")
+	generated, err := Resources(ns, testRepoURL, "")
 	if err != nil {
 		t.Fatal(err)
 	}
@@ -99,7 +88,6 @@ func TestResource(t *testing.T) {
 	sa := roles.CreateServiceAccount(name)
 	want := res.Resources{
 		"02-rolebindings/commit-status-tracker-service-account.yaml": sa,
-		"03-secrets/commit-status-tracker.yaml":                      testSecret,
 		"02-rolebindings/commit-status-tracker-role.yaml":            roles.CreateRole(name, roleRules),
 		"02-rolebindings/commit-status-tracker-rolebinding.yaml":     roles.CreateRoleBinding(name, sa, "Role", operatorName),
 		"10-commit-status-tracker/operator.yaml":                     createStatusTrackerDeployment(ns, "https://github.com/testing/testing.git", ""),

--- a/pkg/pipelines/utils.go
+++ b/pkg/pipelines/utils.go
@@ -1,0 +1,41 @@
+package pipelines
+
+import (
+	"errors"
+	"fmt"
+	"net/url"
+	"os"
+
+	"github.com/mitchellh/go-homedir"
+	"github.com/spf13/afero"
+)
+
+// check if the file exists or not
+// TODO: This should be renamed for clarity.
+func CheckFileExists(fs afero.Fs, dockerConfigJSONFilename string) (string, error) {
+	if dockerConfigJSONFilename == "" {
+		return "", errors.New("failed to generate path to file: must provide --dockerconfigjson parameter")
+	}
+	authJSONPath, err := homedir.Expand(dockerConfigJSONFilename)
+	if err != nil {
+		return "", fmt.Errorf("failed to generate path to file: %v", err)
+	}
+	_, err = fs.Stat(authJSONPath)
+	if os.IsNotExist(err) {
+		return "", fmt.Errorf("file does not exist at the file-path passed to access the dockercfgjson, kindly enter a valid file path")
+	}
+	if err != nil {
+		return "", fmt.Errorf("failed to read Docker config %q : %s", authJSONPath, err)
+	}
+	return authJSONPath, nil
+}
+
+func repoURL(u string) (string, error) {
+	parsed, err := url.Parse(u)
+	if err != nil {
+		return "", fmt.Errorf("failed to parse %q: %w", u, err)
+	}
+	parsed.Path = ""
+	parsed.User = nil
+	return parsed.String(), nil
+}

--- a/pkg/pipelines/utils_test.go
+++ b/pkg/pipelines/utils_test.go
@@ -1,0 +1,26 @@
+package pipelines
+
+import "testing"
+
+func TestRepoURL(t *testing.T) {
+	urlTests := []struct {
+		repoURL string
+		wantURL string
+	}{
+		{"https://github.com/my-org/my-repo.git", "https://github.com"},
+		{"https://gl.example.com/my-org/my-repo.git", "https://gl.example.com"},
+	}
+
+	for _, tt := range urlTests {
+		t.Run(tt.repoURL, func(rt *testing.T) {
+			u, err := repoURL(tt.repoURL)
+			if err != nil {
+				rt.Error(err)
+				return
+			}
+			if u != tt.wantURL {
+				rt.Errorf("got %q, want %q", u, tt.wantURL)
+			}
+		})
+	}
+}

